File: api/src/main/java/gg/mineral/api/entity/attribute/AttributeInstance.java
package gg.mineral.api.entity.attribute;

import java.util.Collection;
import java.util.Map;
import java.util.UUID;
import java.util.function.Consumer;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
import lombok.Getter;
import lombok.val;

/**
 * Represents an instance of an attribute and its modifiers.
 */
public class AttributeInstance {
    @Getter
    private final Attribute attribute;
    private final Map<UUID, AttributeModifier> modifiers = new Object2ObjectOpenHashMap<>();
    private final Consumer<AttributeInstance> propertyChangeListener;
    @Getter
    private float baseValue;
    private float cachedValue = 0.0f;

    public AttributeInstance(@NotNull Attribute attribute, @Nullable Consumer<AttributeInstance> listener) {
        this.attribute = attribute;
        this.propertyChangeListener = listener;
        this.baseValue = attribute.getDefaultValue();
        refreshCachedValue();
    }

    /**
     * Sets the base value of this instance.
     *
     * @param baseValue the new base value
     * @see #getBaseValue()
     */
    public void setBaseValue(float baseValue) {
        if (this.baseValue != baseValue) {
            this.baseValue = baseValue;
            refreshCachedValue();
        }
    }

    /**
     * Add a modifier to this instance.
     *
     * @param modifier the modifier to add
     */
    public void addModifier(@NotNull AttributeModifier modifier) {
        if (modifiers.putIfAbsent(modifier.getId(), modifier) == null)
            refreshCachedValue();
    }

    /**
     * Remove a modifier from this instance.
     *
     * @param modifier the modifier to remove
     */
    public void removeModifier(@NotNull AttributeModifier modifier) {
        if (modifiers.remove(modifier.getId()) != null)
            refreshCachedValue();
    }

    /**
     * Get the modifiers applied to this instance.
     *
     * @return the modifiers.
     */
    @NotNull
    public Collection<AttributeModifier> getModifiers() {
        return modifiers.values();
    }

    /**
     * Gets the value of this instance calculated with modifiers applied.
     *
     * @return the attribute value
     */
    public float getValue() {
        return cachedValue;
    }

    /**
     * Recalculate the value of this attribute instance using the modifiers.
     */
    protected void refreshCachedValue() {
        final Collection<AttributeModifier> modifiers = getModifiers();
        float base = getBaseValue();

        for (val modifier : modifiers.stream().filter(mod -> mod.getOperation() == AttributeOperation.ADDITION)
                .toArray(AttributeModifier[]::new))
            base += modifier.getAmount();

        float result = base;

        for (val modifier : modifiers.stream().filter(mod -> mod.getOperation() == AttributeOperation.MULTIPLY_BASE)
                .toArray(AttributeModifier[]::new))
            result += (base * modifier.getAmount());

        for (val modifier : modifiers.stream().filter(mod -> mod.getOperation() == AttributeOperation.MULTIPLY_TOTAL)
                .toArray(AttributeModifier[]::new))
            result *= (1.0f + modifier.getAmount());

        this.cachedValue = Math.min(result, getAttribute().getMaxValue());

        // Signal entity
        if (propertyChangeListener != null)
            propertyChangeListener.accept(this);
    }
}

File: api/src/main/java/gg/mineral/api/entity/attribute/AttributeInstance.kt
package gg.mineral.api.entity.attribute

import gg.mineral.api.entity.attribute.Attribute.maxValue
import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap
import lombok.Getter
import java.util.*
import java.util.function.Consumer
import kotlin.math.min

/**
 * Represents an instance of an attribute and its modifiers.
 */
class AttributeInstance(
    @field:Getter private val attribute: Attribute,
    private val propertyChangeListener: Consumer<AttributeInstance>?
) {
    private val modifiers: MutableMap<UUID, AttributeModifier?> = Object2ObjectOpenHashMap()

    @Getter
    private var baseValue: Float

    /**
     * Gets the value of this instance calculated with modifiers applied.
     *
     * @return the attribute value
     */
    var value: Float = 0.0f
        private set

    init {
        this.baseValue = attribute.defaultValue
        refreshCachedValue()
    }

    /**
     * Sets the base value of this instance.
     *
     * @param baseValue the new base value
     * @see .getBaseValue
     */
    fun setBaseValue(baseValue: Float) {
        if (this.baseValue != baseValue) {
            this.baseValue = baseValue
            refreshCachedValue()
        }
    }

    /**
     * Add a modifier to this instance.
     *
     * @param modifier the modifier to add
     */
    fun addModifier(modifier: AttributeModifier) {
        if (modifiers.putIfAbsent(modifier.id, modifier) == null) refreshCachedValue()
    }

    /**
     * Remove a modifier from this instance.
     *
     * @param modifier the modifier to remove
     */
    fun removeModifier(modifier: AttributeModifier) {
        if (modifiers.remove(modifier.id) != null) refreshCachedValue()
    }

    /**
     * Get the modifiers applied to this instance.
     *
     * @return the modifiers.
     */
    fun getModifiers(): Collection<AttributeModifier?> {
        return modifiers.values
    }

    /**
     * Recalculate the value of this attribute instance using the modifiers.
     */
    protected fun refreshCachedValue() {
        val modifiers = getModifiers()
        var base: Float = getBaseValue()

        for (modifier in modifiers.stream()
            .filter { mod: AttributeModifier? -> mod!!.operation == AttributeOperation.ADDITION }
            .toArray<AttributeModifier> { _Dummy_.__Array__() }) base += modifier.amount

        var result = base

        for (modifier in modifiers.stream()
            .filter { mod: AttributeModifier? -> mod!!.operation == AttributeOperation.MULTIPLY_BASE }
            .toArray<AttributeModifier> { _Dummy_.__Array__() }) result += (base * modifier.amount)

        for (modifier in modifiers.stream()
            .filter { mod: AttributeModifier? -> mod!!.operation == AttributeOperation.MULTIPLY_TOTAL }
            .toArray<AttributeModifier> { _Dummy_.__Array__() }) result *= (1.0f + modifier.amount)

        this.value = min(result.toDouble(), getAttribute().maxValue.toDouble()).toFloat()

        // Signal entity
        propertyChangeListener?.accept(this)
    }
}